util/topgen.py

```diff
@@ -663,32 +663,53 @@ def generate_top_ral(top: Dict[str, object],
     # Collect up the memories to add
     mems = []
     for item in list(top.get("memory", [])):
-        byte_write = ('byte_write' in item and
-                      item["byte_write"].lower() == "true")
-        data_intg_passthru = ('data_intg_passthru' in item and
-                              item["data_intg_passthru"].lower() == "true")
-        size_in_bytes = int(item['size'], 0)
-        num_regs = size_in_bytes // addrsep
-        swaccess = access.SWAccess('top-level memory',
-                                   item.get('swaccess', 'rw'))
-
-        mems.append(window.Window(name=item['name'],
-                                  desc='(generated from top-level)',
-                                  unusual=False,
-                                  byte_write=byte_write,
-                                  data_intg_passthru=data_intg_passthru,
-                                  validbits=regwidth,
-                                  items=num_regs,
-                                  size_in_bytes=size_in_bytes,
-                                  offset=int(item["base_addr"], 0),
-                                  swaccess=swaccess))
+        offset = int(item['base_addr'], 0)
+        mems.append(create_mem(item, item['name'], addrsep, regwidth, offset))
+
+    # if memory is defined in top['module'], use it to replace the mem in the block
+    for module in top['module']:
+        if "memory" in module:
+            block = name_to_block[module['type']]
+            for mem_name, item in module['memory'].items():
+                assert block.reg_blocks[mem_name]
+                assert len(block.reg_blocks[mem_name].windows) == 1
+
+                if 'base_addr' in item:
+                    offset = int(item['base_addr'], 0)
+                else:
+                    offset = 0
+
+                block.reg_blocks[mem_name].windows[0] = create_mem(
+                    item, mem_name, addrsep, regwidth, offset)
 
     chip = Top(regwidth, name_to_block, inst_to_block, if_addrs, mems, attrs)
 
     # generate the top ral model with template
     return gen_dv(chip, dv_base_prefix, str(out_path))
 
 
+def create_mem(item, name, addrsep, regwidth, offset):
```

Comment:
No need to pass `name` as a separate argument here, I don't think (you can get it from `item['name']`)

Reply:
sorry, I missed this comment.

Since in the top, memory is like this, which is slightly different than the `window` in the block
```
      memory: {
        rom: {
          label:      "rom",
          swaccess:   "ro",
          exec:       "True",
          byte_write: "False",
          size:       "0x4000"
        }
      },
```
item['name'] doesn't exist in this data structure.

I added below and also removed the input `name` in create_mem, so that the structure of `item` is more consistent.
```
            for mem_name, item in module['memory'].items():
                ...
                item['name'] = mem_name
```